dvc/objects/stage.py

```diff
@@ -151,19 +157,37 @@ def _get_tree_obj(path_info, fs, name, odb, state, upload, **kwargs):
     if value:
         hash_info = HashInfo(name, value)
         try:
-            return Tree.load(odb, hash_info)
+            tree = Tree.load(odb, hash_info)
+            # NOTE: loaded entries are naive objects with hash_infos but no
+            # path_info. For staging trees, obj.path_info should be relative
+            # to the staging src `path_info` and src fs
+            for key, entry in tree:
+                entry.fs = fs
+                entry.path_info = path_info.joinpath(*key)
+            return tree
         except FileNotFoundError:
             pass
 
     tree = _build_tree(path_info, fs, name, odb, state, upload, **kwargs)
 
+    if odb.staging is not None:
+        odb = odb.staging
+
     odb.add(tree.path_info, tree.fs, tree.hash_info)
-    if name != "md5":
+
+    raw = odb.get(tree.hash_info)
+    if name == "md5":
+        # cleanup the digest() generated memfs path if needed and return the
+        # staged tree object based on the ODB fs/path
+        if tree.fs.exists(tree.path_info):
+            tree.fs.remove(tree.path_info)
```

Comment:
When staging in memory, the original `tree.path_info` will already be removed (since the `odb.add` will use `fs.move()`). When staging into local, it will be copied from memory into the local staging ODB, and we can explicitly remove the temp memfile here.